### PR TITLE
Address bugs for OData Client Code Gen

### PR DESCRIPTION
**Pull request by AlanWong-MS**
### Issues
#985, #1044, #1148

### Description
This changes enables compatibility with VS2015, addresses the IgnoreUnexpectedAttributesAndElements bug where the setting does not get propagated to the generated code (special thanks to @tchuan for taking a look at it), and adds a logo to the extension.

### Checklist (Uncheck if it is not completed)
- [x] *Test cases added*
- [x] *Build and test with one-click build and test script passed*

### Additional Items
A separate PR to address the .VB side as well. #1197

### CODE REVIEW

**Comment by AlanWong-MS** [comment]
Need to address .VB side as well.

File: src/CodeGen/ODataT4CodeGenerator.cs [hunk 1]
```diff
@@ -3850,7 +3865,23 @@ internal override void WriteGeneratedEdmModel(string escapedEdmxString)
...
+this.Write(this.context.IgnoreUnexpectedElementsAndAttributes.ToString().ToLower());
+this.Write(@"
+                    };
+                    
+                    global::System.Collections.Generic.IEnumerable<global::Microsoft.OData.Edm.Validation.EdmError> errors;
+                    global::System.Collections.Generic.IEnumerable<global::Microsoft.OData.Edm.IEdmModel> references = global::System.Linq.Enumerable.Empty<global::Microsoft.OData.Edm.IEdmModel>();
+                    global::Microsoft.OData.Edm.IEdmModel edmModel;
+                    
+                    if (!global::Microsoft.OData.Edm.Csdl.CsdlReader.TryParse(reader, references, edmxReaderSettings, out edmModel, out errors))
+                    {
```
Comment (mikepizzo) [inline src/CodeGen/ODataT4CodeGenerator.cs]: If all you are setting on readerSettings is IgnoreUnexpectedAttributesAndElements, why not use this overload, and not build readerSettings or references?:
       public static bool TryParse(XmlReader reader, bool ignoreUnexpectedAttributesAndElements, out IEdmModel model, out IEnumerable<EdmError> errors)

File: src/CodeGen/ODataT4CodeGenerator.ttinclude [hunk 2]
```diff
@@ -3491,7 +3514,34 @@ namespace <#= fullNamespace #>
...
+            }
+#>
+                    };
+                    
+                    global::System.Collections.Generic.IEnumerable<global::Microsoft.OData.Edm.Validation.EdmError> errors;
+                    global::System.Collections.Generic.IEnumerable<global::Microsoft.OData.Edm.IEdmModel> references = global::System.Linq.Enumerable.Empty<global::Microsoft.OData.Edm.IEdmModel>();
+                    global::Microsoft.OData.Edm.IEdmModel edmModel;
+                    
+                    if (!global::Microsoft.OData.Edm.Csdl.CsdlReader.TryParse(reader, references, edmxReaderSettings, out edmModel, out errors))
+                    {
```
Comment (mikepizzo) [inline src/CodeGen/ODataT4CodeGenerator.ttinclude]: here, too, should be able to use:
       public static bool TryParse(XmlReader reader, bool ignoreUnexpectedAttributesAndElements, out IEdmModel model, out IEnumerable<EdmError> errors)

File: src/CodeGen/ODataT4CodeGenerator.ttinclude [hunk 3]
```diff
@@ -3473,12 +3473,35 @@ namespace <#= fullNamespace #>
                 global::System.Xml.XmlReader reader = CreateXmlReader(Edmx);
                 try
                 {
-                    return global::Microsoft.OData.Edm.Csdl.CsdlReader.Parse(reader, getReferencedModelFromMap);
-                }
-                finally
```
Comment (mikepizzo) [inline src/CodeGen/ODataT4CodeGenerator.ttinclude]: where did the finally end up?

File: src/CodeGen/ODataT4ItemTemplate/ODataT4ItemTemplate.csproj [hunk 4]
```diff
@@ -57,6 +57,10 @@
       <CopyToOutputDirectory>Always</CopyToOutputDirectory>
       <IncludeInVSIX>true</IncludeInVSIX>
     </Content>
+    <Content Include="odata_logo.png">
```
Comment (xuzhg) [inline src/CodeGen/ODataT4ItemTemplate/ODataT4ItemTemplate.csproj]: >odata_logo.png [](start = 22, length = 14)

That's a missing file?

File: src/CodeGen/ODataT4CodeGenerator.cs [hunk 5]
```diff
@@ -3850,7 +3865,23 @@ internal override void WriteGeneratedEdmModel(string escapedEdmxString)
...
+                        throw new global::System.InvalidOperationException(""Error while parsing CSDL."");
```
Comment (xuzhg) [inline src/CodeGen/ODataT4CodeGenerator.cs]: >global::System.InvalidOperationException(""Error while parsing CSDL."") [](start = 34, length = 71)

"Error while parsing CSDL." looks meaningless.
Maybe better to include the errors into the exception?

File: src/CodeGen/ODataT4ItemTemplate/ODataT4ItemTemplate.csproj (hunk 4 above)
Comment (AlanWong-MS) [reply src/CodeGen/ODataT4ItemTemplate/ODataT4ItemTemplate.csproj]: It's added in this PR. It's the file listed right below.

File: src/CodeGen/ODataT4CodeGenerator.ttinclude (hunk 3 above)
Comment (AlanWong-MS) [reply src/CodeGen/ODataT4CodeGenerator.ttinclude]: Thanks for the catch. Finally, finally ended up back here.

File: src/CodeGen/ODataT4CodeGenerator.cs (hunk 5 above)
Comment (AlanWong-MS) [reply src/CodeGen/ODataT4CodeGenerator.cs]: Thanks, changed to:

```
if (!global::Microsoft.OData.Edm.Csdl.CsdlReader.TryParse(reader, references, edmxReaderSettings, out edmModel, out errors))
{
    global::System.Collections.Generic.List<Microsoft.OData.Edm.Validation.EdmError> errorMessages = new global::System.Collections.Generic.List<Microsoft.OData.Edm.Validation.EdmError>(errors);
    throw new global::System.InvalidOperationException(string.Join(";", errorMessages.ConvertAll(e => e.ErrorMessage).ToArray()));
}
```

File: src/CodeGen/ODataT4CodeGenerator.ttinclude (hunk 2 above)
Comment (AlanWong-MS) [reply src/CodeGen/ODataT4CodeGenerator.ttinclude]: :+1:

File: src/CodeGen/ODataT4CodeGenerator.cs (hunk 1 above)
Comment (AlanWong-MS) [reply src/CodeGen/ODataT4CodeGenerator.cs]: Thanks for pointing that out! This greatly simplifies the code.

File: src/CodeGen/ODataT4CodeGenerator.cs [hunk 6]
```diff
@@ -3850,7 +3848,23 @@ internal override void WriteGeneratedEdmModel(string escapedEdmxString)
...
+                    this.Write(this.context.IgnoreUnexpectedElementsAndAttributes.ToString().ToLower());
+                    this.Write(@", out edmModel, out errors))
+                    {
+                        global::System.Text.StringBuilder errorMessages = new System.Text.StringBuilder();
+                        foreach (var error in errors)
+                        {
+                            errorMessages.Append(error.ErrorMessage);
+                            errorMessages.Append(""; "");
+                        }
+                        throw new global::System.InvalidOperationException(errorMessages.ToString());
```
Comment (AlanWong-MS) [inline src/CodeGen/ODataT4CodeGenerator.cs]: note: I used a LINQ expression to concatenate the error messages at first but one of the generated files didn't seem to like that so I changed it to this foreach loop.

Comment (mikepizzo) [reply src/CodeGen/ODataT4CodeGenerator.cs]: Is that change not part of this PR?

---
In reply to: [197884091](https://github.com/OData/odata.net/pull/1150#discussion_r197884091) [](ancestors = 197884091)

**Pull request merged**